enable empty batch for all flavors of convolution

ngimel opened the pull request:
Enables convolution with an empty batch or number of channels for all flavors of convolution (grouped convolution, convTranspose). Would make #31658 unnecessary. 
There's a philosophical question of whether we need to return zero gradients for weights if weights did not materially participate in the computation, or is it ok to leave weight gradients as None. I vote for the latter, the former is not easy to achieve without modifying code in all the backends, which is what #31658 does, and #26214 did before it, but it also does not add much value. If we decide that returning zero gradients for weights is important, I still think it should be done in the backend-independent manner, by having backend-independent empty_conv_forward function with its defined backward. A strike against #26214 approach is that it enables (meaningless) double backward through weights, ok, but any further backwards will break because returned gradients for double-backward are `zeros_like`, disconnected from the graph, so composability ends at double backward.
I'm returning a view of empty input, I'm open to the idea that it should be a clone, I also don't see how it matters if the input is empty anyway ;-)

Commit:
enable empty batch for all flavors of convolution

Commit:
enable empty batch for all flavors of convolution

Commit:
add test for convTranspose

Commit:
add test for convTranspose

Commit:
lint

Commit:
lint

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit d8bc2bcb:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d8bc2bcb8656c08604edafe5cb102b37256ff8c7).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4340126) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90693478) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90693478))
<details>
<summary>
<code>Jan 24 22:04:00 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 24 22:03:59 Ran 1047 tests in 264.011s 
Jan 24 22:03:59  
Jan 24 22:03:59 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 24 22:03:59  
Jan 24 22:03:59 Generating XML reports... 
Jan 24 22:04:00 Traceback (most recent call last): 
Jan 24 22:04:00   File "test/run_test.py", line 456, in <module> 
Jan 24 22:04:00     main() 
Jan 24 22:04:00   File "test/run_test.py", line 449, in main 
Jan 24 22:04:00     raise RuntimeError(message) 
Jan 24 22:04:00 RuntimeError: test_autograd failed! 
Jan 24 22:04:00 + cleanup 
Jan 24 22:04:00 + retcode=1 
Jan 24 22:04:00 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ngimel commented:
Test failures are real, but are related to not returning weight gradients, which, I maintain, should not be returned in this case.

xush6528 committed:
Support TorchScript call over remote API (RRef) (#32466)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32466

It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

ghstack-source-id: 97211168

Test Plan:
# Unit tests

```
buck test mode/dev-nosan //caffe2/test/distributed/rpc:rpc_fork -- test_script_function_exception

buck build mode/dev-nosan //caffe2/test/distributed/rpc:rpc_fork

buck-out/gen/caffe2/test/distributed/rpc/rpc_fork\#binary.par -r test_script_function_exception
```

```
buck test mode/dev-nosan //caffe2/test/distributed/rpc:dist_autograd_fork -- test_backward_simple_script_call

buck build mode/dev-nosan //caffe2/test/distributed/rpc:dist_autograd_fork

buck-out/gen/caffe2/test/distributed/rpc/dist_autograd_fork\#binary.par -r test_backward_simple_script_call
```

Differential Revision: D19440633

fbshipit-source-id: d37f6dcdc0b80d35ac7bcba46ad6f9b831c3779b

jspark1105 committed:
[caffe2] use JIT'ed fp32 SLS (#32413)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32413

Use JIT'ed fp32 SLS in Caffe2 operators

Test Plan: CI

Reviewed By: jianyuh

Differential Revision: D19460555

fbshipit-source-id: 4f29d34523efb6ea1e4c324cc8c93c96990c6aad

ljk53 committed:
[pytorch] move type_derived_methods out of anonymous namespace (#32275)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32275

Currently TypeDerived (e.g. `CPUType::`) methods are declared and
defined in anonymous namespace as they are only called from c10
dispatcher - except for STATIC_DISPATCH mode, where they can be directly
called from Functions.h.

We plan to generate c10 op registration into separate files for internal
xplat/BUCK build, thus we need declare these methods in non-anonymous
namespace.

I feel it's easier to simply change it unconditionally, unless there are
some side effect I'm not aware of - `TypeDefault::` methods are in
non-anonymous namespace anyway.
ghstack-source-id: 97214789

Test Plan: - CI

Differential Revision: D19426692

Pulled By: ljk53

fbshipit-source-id: 44aebba15f5e88ef4acfb623844f61d735016959

ljk53 committed:
[pytorch] codegen flags to whitelist op registrations / generate to separate files (#32451)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32451

This PR adds a few new parameters to ATen codegen script:

```
1. op_registration_whitelist
Can be used to filter op registrations for selective build;

2. type_whitelist
Can be used to filter types (CPUType, CUDAType, ...) for selective build;

3. per_op_registration
When set it will group function registrations by op name and write to separate files;
```

1 & 2 are introduced for mobile custom build without relying on static dispatch;
3 is introduced to solve custom build with multi-library / multi-model (needed by FB
internal build - see more details: https://fb.quip.com/ZVh1AgOKW8Vv).

These flags should work independently with each other (and independent to USE_STATIC_DISPATCH).

Not setting them should have no effect compared to master.
ghstack-source-id: 97214788

Test Plan: - tested all 3 params with FB internal build changes.

Differential Revision: D19427919

fbshipit-source-id: a381fe5f768fe2e9196563787f08eb9f18316e83

jeongukjae committed:
Fix wrong typing (torch/nn/parameter.pyi) (#32617)

Summary:
A constructor of `nn.Parameter` has default values on `data` and `requires_grad`, but in type stub, there are no default values.

Resolve https://github.com/pytorch/pytorch/issues/32481
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32617

Differential Revision: D19571397

Pulled By: ngimel

fbshipit-source-id: fd14298aa472b7575221229cecf5a56f8c84f531

Commit:
Fix MagmaInitializesCorrectly_CUDA by using an invertible matrix (#32547)

Summary:
This test case had been using the tensor

```
1  2  3  4
5  6  7  8
9  10 11 12
13 14 15 16
```

which is not an invertible tensor and causes the test case to fail, even if magma gets initialized just fine. This change uses a tensor that is invertible, and the inverse doesn't include any elements that are close to zero to avoid floating point rounding errors.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32547

Differential Revision: D19572316

Pulled By: ngimel

fbshipit-source-id: 1baf3f8601b2ba69fdd6678d7a3d86772d01edbe

Enealor committed:
Update linspace types (#32218)

Summary:
Changes the linspace functions to be more consistent as requested in https://github.com/pytorch/pytorch/issues/31991. The code has also been updated to avoid an early rounding error; the line `scalar_t step = (scalar_end - scalar_start) / static_cast<static_t>(steps-1)` can result in `step = 0` for integer scalars, and this gives unintended results. I examined the new output using
```
import torch

types = [torch.uint8, torch.int8, torch.short, torch.int, torch.long, torch.half, torch.float, torch.double]

print('Testing linspace:')
for type in types:
    print(type, torch.linspace(-2, 2, 10, dtype=type))
```
which returns
```
Testing linspace:
torch.uint8 tensor([254, 254, 254, 255, 255,   0,   0,   1,   1,   2], dtype=torch.uint8)
torch.int8 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2], dtype=torch.int8)
torch.int16 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2], dtype=torch.int16)
torch.int32 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2], dtype=torch.int32)
torch.int64 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2])
torch.float16 tensor([-2.0000, -1.5557, -1.1113, -0.6670, -0.2227,  0.2227,  0.6660,  1.1113,
         1.5547,  2.0000], dtype=torch.float16)
torch.float32 tensor([-2.0000, -1.5556, -1.1111, -0.6667, -0.2222,  0.2222,  0.6667,  1.1111,
         1.5556,  2.0000])
torch.float64 tensor([-2.0000, -1.5556, -1.1111, -0.6667, -0.2222,  0.2222,  0.6667,  1.1111,
         1.5556,  2.0000], dtype=torch.float64)
```
which is the expected output: `uint8` overflows as it should, and the result of casting from a floating point to an integer is correct.

This PR does not change the logspace function.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32218

Differential Revision: D19544224

Pulled By: ngimel

fbshipit-source-id: 2bbf2b8552900eaef2dcc41b6464fc39bec22e0b

andfoy committed:
Add warning regarding pickle insecurity on torch.load documentation (#32593)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/31875

Added a small warning box based on the one presented on the [pickle](https://docs.python.org/3/library/pickle.html) module regarding the safety issues of unpickling files. i.e., unwanted code execution.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32593

Differential Revision: D19572292

Pulled By: ngimel

fbshipit-source-id: 69e7de390133ea77bddcadcd5b6820193c8abcc9

csummersea committed:
Bug fix of norm minimization for dev mode (#31462)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31462

Fix the divide by zero issue in norm minimization in dev mode

Test Plan: buck run mode/dev vision/video_modeling/classification/tools:test_octGloRe_quantization -- --test_data=/mnt/vol/gfsadslearner-frc3c01/fblearner_flow/users/summerdeng/xray_video/deep_vision_video_yufei_test_data_fcc_v4p2_10.csv --output_dir /mnt/vol/gfsadslearner-frc3c01/fblearner_flow/users/summerdeng/xray_video/octGloRe --load_model_path=/mnt/vol/gfsfblearner-oregon/flow/data/2019-10-15/e2681db8-e4f5-4b70-ae18-45bf0b8fbfbc/train_model_epoch0_inputcount0_final.mdl --dataset_name="FCC V4P2" --num_labels=1099 --column_handle="handle" --clip_per_video=1 --num_groups=24 --width_per_group=2 --batch_size=32 --histogram_file=/mnt/vol/gfsadslearner-frc3c01/fblearner_flow/users/summerdeng/xray_video/octGloRe/hist_octGloRe_final_24x2_fcc_v4p2_1clip_f144586257_nullfix_100k_compiled.hist --int8_model_type="pb"  --int8_predict_net_path="reproduce_octGloRe_final_24x2_predict_net_int8_l2approx_wminmax_from_mdl.pb" --int8_init_net_path="reproduce_octGloRe_final_24x2_init_net_int8_l2approx_wminmax_from_mdl.pb" --weight_quant="l2_approx" --activation_quant="l2_approx"  --print_model --int8_model_saved --num_iter 10

Reviewed By: jspark1105

Differential Revision: D19172591

fbshipit-source-id: 994a20e3364b0dc33623a11281e0bdbc2e06159d

Krovatkin committed:
faster bailout tests (#32266)

Summary:
Reduces the overhead of `prim::BailOut` nodes.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32266

Differential Revision: D19503336

Pulled By: Krovatkin

fbshipit-source-id: daa0c373f0fa17edd689600b75e7e4ba98b4670a

peterjc123 committed:
Make cuda search process of cpp extension quiet (#32620)

Summary:
Fixes https://discuss.pytorch.org/t/error-with-cpp-extentions/67559.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32620

Differential Revision: D19576164

Pulled By: soumith

fbshipit-source-id: 076229322375774bec03ef2632fc233000c15391

rohan-varma committed:
[rpc] make handling of FORWARD_AUTOGRAD_REQ in request_callback_impl (#32476)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32476

This makes the handling of FORWARD_AUTOGRAD_REQ in request_callback
nonblocking. Processing this message requires unwrapping the message with
autograd information, processing the original message, and sending back the
message with autograd information wrapped. This makes the processing the
original message nonblocking by grabbing a future to it and marking the parent
future as completed when this one completes.
ghstack-source-id: 97221251

Test Plan: `test_rpc_spawn.py` and `test_dist_autograd_spawn.py` both pass.

Differential Revision: D19509501

fbshipit-source-id: 84ad2f9c5305ed11ed9bb0144b1aaf5f8698cd2b

ezyang commented:
> I'm returning a view of empty input, I'm open to the idea that it should be a clone, I also don't see how it matters if the input is empty anyway ;-)

I think it should be a clone, so that a subsequent `resize_` doesn't affect the original tensor.

ezyang commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -577,10 +565,17 @@ at::Tensor _convolution(
+    weight_sizes_mkl[0] = weight.size(0) * weight.size(1);
+    std::copy_n(
+        weight.sizes().cbegin() + 2, k - 1, weight_sizes_mkl.begin() + 1);
+    weight_sizes = c10::IntArrayRef(weight_sizes_mkl);
```
**WOOP WOOP OWNERSHIP POLICE**. IntArrayRef is a NON-OWNING reference to the pointed to array. You only allocated `weight_sizes_mkl` inside this conditional, which means it goes dead when you exit!

ngimel replied:
oops, thanks!

ezyang reviewed (changes requested):
I don't think viewing the input tensor is appropriate.  If I subsequently `resize_` on the empty tensor, if I have a view on the input tensor, the resize will now expand to cover the old memory from the input tensor. That is a very observable difference!

Commit:
return clone of input, move weight_sizes_mkl to appropriate scope

Commit:
return clone of input, move weight_sizes_mkl to appropriate scope

lcskrishna committed:
[ROCm] Enable Caffe2 video operators for ROCm

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32610

Differential Revision: D19580129

Pulled By: ezyang

fbshipit-source-id: 16d620173dcc231068e041d599aa09c94e677a9e

ngimel commented:
I've fixed scope for weight_sizes_mkl and am returning a clone. What do you think about weight gradients? I have a world ugliest workaround to automatically return infinitely differentiable zero gradients for parameters, but I'd prefer not to use it unless we decide it's necessary. @ppwwyyxx seems to want zero gradients [here](https://github.com/pytorch/pytorch/issues/12013#issuecomment-569876108)

XiaobingSuper committed:
Enable mkldnn on windows (#31355)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/15982.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31355

Differential Revision: D19428979

Pulled By: ezyang

fbshipit-source-id: bee304c5913e70e8dead3098e9796051861cd666

v0dro committed:
verify input sizes for instance norm and group norm (#29082)

Summary:
Fix for https://github.com/pytorch/pytorch/issues/19250
Pull Request resolved: https://github.com/pytorch/pytorch/pull/29082

Differential Revision: D19373507

Pulled By: ezyang

fbshipit-source-id: 231a79280f4cd7db2c26218a60869356a124bf77

v0dro committed:
0-dim batch size input for interpolate. (#32400)

Summary:
This PR adds support for 0-dim batch size input for `torch.nn.functional.interpolate` for various modes of interpolation.

Fixes part of gh-12013

CC: rgommers  ezyang
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32400

Differential Revision: D19557090

Pulled By: ezyang

fbshipit-source-id: 6822f148bb47bfbcacb5e03798bf2744f24a2a32

iotamudelta committed:
[ROCm] Adjust elementwise_kernel settings on ROCm (#32609)

Summary:
Recent PR https://github.com/pytorch/pytorch/issues/31974 and upcoming PR https://github.com/pytorch/pytorch/issues/32383 are changing the behavior of the elementwise_kernel infrastructure on CUDA.

In order to stay in sync, change the nd-loop behavior to match ROCm and CUDA for now. Once the full rework is done, the ROCm settings will likely diverge again.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32609

Differential Revision: D19580121

Pulled By: ezyang

fbshipit-source-id: 4c8dcf6db3ac973e48ece6a665615cfe7d7cb764

Commit:
Fixes moving after weight norm application (#32563)

Summary:
This PR updates how RNNs handle their "flat weights." In particular, it allows for only some flat weights to be "materialized" when apply is called, and it updates the flattening behavior to only apply if all flat weights are (1) materialized, (2) share a dtype and (3) are acceptable to cuDNN.

One test is modified and another created to test these changes. One practical effect of this change is that weight norm can be successfully applied to a module BEFORE that module is moved to an accelerator. Previously doing so would throw an error.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32563

Differential Revision: D19562258

Pulled By: mruberry

fbshipit-source-id: 4fef006e32cdfd8e3e3d519fc2ab5fc203dd7b36

Commit:
Back out "Use simd version for fp16 conversions" (#32640)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32640

Original commit changeset: 3b1ee0ba756e

Reverting according to https://our.intern.facebook.com/intern/diff/D19291499/?transaction_id=1347995678706116&dest_fbid=465672071047258

Test Plan: unittests.

Reviewed By: jspark1105, jianyuh

Differential Revision: D19576708

fbshipit-source-id: bec92318523498067935234ab702c925ece71da6

ezyang committed:
Revert D19562258: [pytorch][PR] Fixes moving after weight norm application

Test Plan: revert-hammer

Differential Revision:
D19562258

Original commit changeset: 4fef006e32cd

fbshipit-source-id: 62e40de19331a61f4a65b7371460fe7dc28f23ea

Commit:
Adding native qconcat

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32252

Test Plan: Imported from OSS

Differential Revision: D19422889

Pulled By: z-a-f

fbshipit-source-id: 23dd5f50009cc4c46b36c39ae1168b57f9a977a4

gchanan committed:
Kill THIntegerTensor, THDenseTensor, THDenseIndexTensor. (#32599)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32599

these aren't used anymore.

Test Plan: Imported from OSS

Differential Revision: D19565655

Pulled By: gchanan

fbshipit-source-id: c0da31365df7342352f9850ae2a2e1e611a6886b

ezyang committed:
Revert "Fix iterator for ncclCommWatchdog. (#32571)" (#32649)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32649

This reverts commit 59dbece3716775c3e6f3a428f73fbf1bde8fac4f.

Revert "Enhance NCCL watchdog to acitvely abort communicators for timed out ops. (#32338)"

This reverts commit f86d6c6afd0e981642d20b4269837334ec46c140.

Test Plan: Imported from OSS

Differential Revision: D19584224

Pulled By: ezyang

fbshipit-source-id: 6cc0ad56ba1f3aec5b48db44e8c6c24c8105db4a

Commit:
Updating submodules

Summary:
GitHub commits:

https://github.com/facebook/fbthrift/commit/9ae8cbb0a1ee7ec2ec6f520ca9cc3c4e884daa4c
https://github.com/facebook/rocksdb/commit/986df37135214d91e543f43a14d32a228b8d6fbb
https://github.com/pytorch/fbgemm/commit/ef4d11b6e125dd09f5d69f6fc999454a516dab6e

Test Plan: n/a

Reviewed By: 2d2d2d2d2d

fbshipit-source-id: 04e7a5ad02cb412ef36672ec30e10a898c525232

Commit:
set up history for weights in case of empty inputs

Commit:
set up history for weights in case of empty inputs

Commit:
[quantization] FP16 dynamic quantized Linear

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32331

Test Plan: Imported from OSS

Differential Revision: D19441158

Pulled By: jamesr66a

fbshipit-source-id: c04247ffe707be68718c486c31bc6c6040f7dc11

IvanKobzarev committed:
[vulkan][caffe2] Add logging for descriptor extensions, fp16 storage

Summary:
`fbcode/caffe2/caffe2/mobile/contrib/libvulkan-stub/BUCK` changes comment:

libvulkan-stub contains vulkan headers `VK_HEADER_VERSION 29`

fbandroid uses ndk r17 that includes vulkan `VK_HEADER_VERSION 76`
which contains defines for extensions that we need.

`("include", "**/*.h"),` -> `("include", "*.h"),` means that ndk vulkan headers to use.

For fp16_storage logging need to add boilerplate for `vkGetPhysicalDeviceFeatures2KHR`

Test Plan:
scuba employees device_event

logcat getVulkanInfo().
```
instance ext.name:VK_KHR_surface
instance ext.name:VK_KHR_android_surface
instance ext.name:VK_EXT_swapchain_colorspace
instance ext.name:VK_KHR_get_surface_capabilities2
instance ext.name:VK_EXT_debug_report
instance ext.name:VK_KHR_device_group_creation
instance ext.name:VK_KHR_external_fence_capabilities
instance ext.name:VK_KHR_external_memory_capabilities
instance ext.name:VK_KHR_get_physical_device_properties2
instance ext.name:VK_KHR_external_semaphore_capabilities
device ext.name:VK_KHR_incremental_present
device ext.name:VK_EXT_hdr_metadata
device ext.name:VK_KHR_shared_presentable_image
device ext.name:VK_GOOGLE_display_timing
device ext.name:VK_KHR_push_descriptor
device ext.name:VK_KHR_image_format_list
device ext.name:VK_EXT_queue_family_foreign
device ext.name:VK_ANDROID_external_memory_android_hardware_buffer
device ext.name:VK_KHR_external_semaphore_fd
device ext.name:VK_KHR_external_fence_fd
device ext.name:VK_KHR_external_memory_fd
device ext.name:VK_KHR_external_memory
device ext.name:VK_KHR_swapchain
device ext.name:VK_KHR_external_semaphore
device ext.name:VK_KHR_driver_properties
device ext.name:VK_KHR_sampler_mirror_clamp_to_edge
device ext.name:VK_KHR_multiview
device ext.name:VK_KHR_relaxed_block_layout
device ext.name:VK_KHR_maintenance1
device ext.name:VK_KHR_maintenance3
device ext.name:VK_KHR_maintenance2
device ext.name:VK_EXT_global_priority
device ext.name:VK_KHR_get_memory_requirements2
device ext.name:VK_KHR_descriptor_update_template
device ext.name:VK_KHR_bind_memory2
device ext.name:VK_KHR_shader_draw_parameters
device ext.name:VK_KHR_dedicated_allocation
device ext.name:VK_KHR_create_renderpass2
device ext.name:VK_KHR_draw_indirect_count
device ext.name:VK_KHR_sampler_ycbcr_conversion
device ext.name:VK_KHR_device_group
device ext.name:VK_KHR_external_fence
device ext.name:VK_KHR_variable_pointers
device ext.name:VK_EXT_sampler_filter_minmax
device ext.name:VK_KHR_storage_buffer_storage_class
VULKAN_SYMBOL_WRAPPER_LOAD_INSTANCE_SYMBOL(vkGetPhysicalDeviceFeatures2KHR) res=1
mChipsetInfoUtilInfo.getVulkanInfo():{vk_driver_version=2149056512, vk_device_id=100859905, vk_extension_descriptor_update_template=1, vk_api_version=4198487, vk_support_fp16_storage=0, vk_platform_dlopen=success, vk_shader_int16=1, vk_device_type=1, vk_shader_float64=0, vk_extension_push_descriptor=1, vk_shader_int64=0, vk_wrapper_init=true, vk_vendor_id=20803, vk_max_compute_shared_memory_size=32768, vk_device_name=Adreno (TM) 630, vk_max_compute_work_group_invocations=1024, vk_device_count=1}
```

Reviewed By: dreiss

Differential Revision: D19564664

fbshipit-source-id: 908b34bdcc24d9b03ecc185edbc5cfb6e7aa27c9

xush6528 committed:
Make _wait_all_workers() support being called for multiple times (#32624)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32624

We need this PR to resolve the issue mentioned in https://github.com/pytorch/pytorch/issues/31325#issuecomment-574918917.

The solution is for each `_wait_all_workers()` call, there is a sequence ID added, to identify different calls.
ghstack-source-id: 97277591

Test Plan:
```
buck test mode/dev-nosan //caffe2/test/distributed/rpc:rpc_fork -- test_wait_all_workers

buck build mode/dev-nosan //caffe2/test/distributed/rpc:rpc_fork

buck-out/gen/caffe2/test/distributed/rpc/rpc_fork\#binary.par -r test_wait_all_workers
```

Differential Revision: D5739520

fbshipit-source-id: a64131e09c365179624700514422f5375afe803f

ppwwyyxx commented:
My main motivation for using zero gradients is to make DDP happy. DDP currently cannot sync gradients properly if `p.grad` is None in some but not all workers. However, alternatives such as `horovod.DistributedOptimizer` seems to handle such cases well. As a result we use hacks such as https://github.com/facebookresearch/detectron2/blob/2292cb3f24cea3d5f77f36dbd8aaf1b8d8d14e08/detectron2/layers/wrappers.py#L101-L104 to please DDP.

If the above issue is solved, then I'm not against None gradients. The only issue I see with it is inconsistency, or a lack of precise semantics -- since it's not clear when None should be used and when zeros should be used. For example, currently `(x+y).sum().backward()` where `x.shape=[0,3], y.shape=[1,3]` will produce zero gradients for y.

albanD reviewed (commented):
>  I'm returning a view of empty input, I'm open to the idea that it should be a clone, I also don't see how it matters if the input is empty anyway ;-)

Autograd inplace logic 🎉 
But given that you discard the base of the view, it should not be a problem.


Given that this is a linear operation. Could a formula like this be simpler (given that you do reduction on the weights anyway):
```
out = (inp.sum() * w.sum() + b.sum()).view(1, 1, 1, 1).expand(output_size)
```

ngimel commented:
Yeah I was thinking about simplifying the formula, and that should work (especially given that otherwise I'd have to add special casing for group convolutions that don't work yet. Thanks!

Commit:
simplify output formula, fix tests

Commit:
merge

ngimel commented:
Closing, a merge gone wrong.
